src/network/networkd-link.c

```diff
@@ -2024,6 +2024,85 @@ static int link_update_master(Link *link, sd_netlink_message *message) {
         return 0;
 }
 
+static int link_update_driver(Link *link, sd_netlink_message *message) {
+        int r;
+
+        assert(link);
+        assert(link->manager);
+        assert(message);
+
+        /* Driver is already read. Assuming the driver is never changed. */
+        if (link->driver)
+                return 0;
+
+        /* When udevd is running, read the driver after the interface is initialized by udevd.
+         * Otherwise, ethtool may not work correctly. See issue #22538.
+         * When udevd is not running, read the value when the interface is detected. */
+        if (link->state != (path_is_read_only_fs("/sys") > 0 ? LINK_STATE_PENDING : LINK_STATE_INITIALIZED))
```

Comment:
given you need this multiple times, and the connection between /sys/ and udev isn't obvious (i mean you added a comment for a reason), maybe wrap the check in a trivial helper function somewhere:

```c
static bool udev_available(void) {
        static int cache = -1;
        if (cache >= 0)
                return cache; 

        cache = path_is_read_only_fs("/sys/") <= 0;
}
```

(and then add a comment there explaining it)

Reply:
Sounds nice!